# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/generic/nsFrameSelection.cpp
# Commit: 8b92fb4e3f41
# Full Hash: 8b92fb4e3f41d856629818bd2ecfc25bed23a643
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2023-12-23 09:52:03
# Regressor Bug: 779684
# File Overlap Count: 1
# Description:
#   Bug 779684 - part 3: Make `nsFrameIterator` non-refcountable r=emilio
#   
#   Nobody shares an instance of `nsFrameIterator`.  Therefore, we can make it
#   non-refcountable.  Additionally, `nsVisualIterator` is not required because
#   it overrides only 4 methods which are really simple.  So, once we merge it
# ==============================================================================

diff -r f357b2287591 -r 8b92fb4e3f41 layout/generic/nsFrameSelection.cpp
--- a/layout/generic/nsFrameSelection.cpp	Sat Dec 23 05:39:38 2023 +0000
+++ b/layout/generic/nsFrameSelection.cpp	Sat Dec 23 05:39:39 2023 +0000
@@ -1024,18 +1024,19 @@
       mozilla::intl::BidiEmbeddingLevel::LTR();
   nsIFrame* foundFrame = aFrameIn;
 
-  RefPtr<nsFrameIterator> frameIterator = nsFrameIterator::Create(
-      mPresShell->GetPresContext(), aFrameIn, nsFrameIterator::Type::Leaf,
-      false,  // aVisual
-      false,  // aLockInScrollView
-      false,  // aFollowOOFs
-      false   // aSkipPopupChecks
+  nsFrameIterator frameIterator(mPresShell->GetPresContext(), aFrameIn,
+                                nsFrameIterator::Type::Leaf,
+                                false,  // aVisual
+                                false,  // aLockInScrollView
+                                false,  // aFollowOOFs
+                                false   // aSkipPopupChecks
   );
-  MOZ_ASSERT(frameIterator);
   do {
     *aFrameOut = foundFrame;
-    foundFrame = frameIterator->Traverse(aDirection == eDirNext);
-    if (!foundFrame) return NS_ERROR_FAILURE;
+    foundFrame = frameIterator.Traverse(aDirection == eDirNext);
+    if (!foundFrame) {
+      return NS_ERROR_FAILURE;
+    }
     foundLevel = foundFrame->GetEmbeddingLevel();
 
   } while (foundLevel > aBidiLevel);